src/py2spack/loading.py

```diff
@@ -0,0 +1,266 @@
+        # TODO: in case of an error, skip the file or return the error?
+        files_parsed: List[Dict[str, str]] = []
+        for f in files_known_format:
+            filename = f["filename"]
+            archive_ext = _parse_archive_extension(filename)
+            if isinstance(archive_ext, APIError):
+                continue
+
+            v = _parse_version(filename, name, archive_ext)
+
+            if v is None:
+                continue
+
+            try:
+                sha256 = f["hashes"]["sha256"]
+            except KeyError:
+                continue
+
+            files_parsed.append(
+                {
+                    "filename": filename,
+                    "url": f["url"],
+                    "version": v,
+                    "extension": archive_ext,
+                    "hash": sha256,
+                }
+            )
+
+        if not files_parsed:
+            return APIError("No valid files found")
+
+        # if argument is -1, return all versions
+        if last_n_versions == -1:
+            return files_parsed
+
+        return files_parsed[-last_n_versions:]
+
+
+def _parse_archive_extension(filename: str) -> str | APIError:
+    extension_list = [
+        ext for ext in KNOWN_ARCHIVE_FORMATS if filename.endswith(ext)
+    ]
+
+    if len(extension_list) == 0:
+        # we return an API error here because the filenames are obtained through
+        # the API and the function is used during the API lookup process
+        msg = f"Extension not recognized for: {filename}"
+        return APIError(msg)
+
+    if len(extension_list) == 1:
+        return extension_list[0]
+    # get the longest matching extension, e.g. .tar.gz instead of .gz
+    longest_matching_ext = max(extension_list, key=len)
+    return longest_matching_ext
+
+
+def _parse_version(
+    filename: str, pkg_name: str, archive_ext: str
+) -> vn.Version | None:
+    """Parse version from filename and check correct formatting."""
+    prefix = pkg_name + "-"
+    if not (filename.startswith(prefix) and filename.endswith(archive_ext)):
+        return None
+    version_str = filename[len(prefix) : -len(archive_ext)]
+    try:
+        parsed_version: vn.Version = vn.parse(version_str)
+        return parsed_version
+
+    except vn.InvalidVersion:
+        return None
+
+
+def _known_archive_format(filename: str):
+    return any([filename.endswith(ext) for ext in KNOWN_ARCHIVE_FORMATS])
+
+
+def _download_sdist(url: str) -> io.BytesIO | APIError:
+    """Download source distribution from url as BytesIO object (in memory)."""
+    response = requests.get(url)
+    if response.status_code == 200:
+        file_like_object = io.BytesIO(response.content)
+        return file_like_object
+    else:
+        filename = url.split("/")[-1]
+        msg = (
+            f"Failed to download sdist {filename},"
+            f" status code: {response.status_code}"
+        )
+        return APIError(msg)
+
+
+def _extract_from_tar(
```

Comment:
This function looks like a general utility without specific business logic. I will move it to some utility module (e.g. `utils.py`)